src/Pact/Typechecker.hs

```diff
@@ -370,21 +371,55 @@ substAppDefun :: Maybe (TcId, AST Node) -> Visitor TC Node
 substAppDefun (Just (defArg,appAst)) Pre t@Var {..}
   | defArg == _aId _aNode = assocAST defArg appAst >> return appAst
   | otherwise = return t
-substAppDefun Nothing Post (App appNode appFun appArgs) = do -- Post, to allow AST subs first
-    af <- case appFun of
-      FNative {} -> return appFun -- noop
-      FDefun {..} -> do
-        -- assemble substitutions, and also associate fun ty with app args
-        let mangledFty = mangleFunType (_aId appNode) _fType
-        subArgs <- forM (zip3 _fArgs appArgs (_ftArgs mangledFty)) $ \(fa,aa,_ft) -> -- do
-          -- assocAstTy (_aNode aa) (_aType ft) -- this was causing recursive var refs, bailing for now
-          return (_aId (_nnNamed fa),aa)
-        -- associate fun return ty with app
-        assocAstTy appNode (_ftReturn mangledFty)
-        let subDefArg b fa = walkAST (substAppDefun (Just fa)) b
-        fb' <- forM _fBody $ \bAst -> foldM subDefArg bAst subArgs
-        return $ set fBody fb' appFun
-    return (App appNode af appArgs)
+substAppDefun Nothing Post (App appNode fun args) = do -- Post, to allow AST subs first
+    fun' <- case fun of
+      FNative {} -> return fun -- noop
+      FDefun {_fType,_fArgs,_fBody} -> do
+        let FunType argTys retTy = mangleFunType (_aId appNode) _fType
+
+        -- associate nodes with types
+
+        -- TODO(stu): this was causing recursive var refs, bailing for now
+        -- for_ (zip args argTys) $ \(arg, argTy) ->
+        --   assocAstTy (_aNode arg) (_aType argTy)
+
```

Comment:
I assume we need the association though, no? Was this replaced by something else?

Reply:
This is carry-over from code you had commented: https://github.com/kadena-io/pact/blob/d4f96422d52464aa5a20b58d4599d13fcc161b76/src/Pact/Typechecker.hs#L380

Do you know what you were referring to at the time?

Reply:
Oh I see. OK just remove it, the only reason I left it commented out was insecurity that it was still going to be necessary.